File: Semester1/Test3/3.2.2/3.2.2/queue.cpp
#include "pch.h"
#include <iostream>
#include "queue.h"
using namespace std;

struct Element
{
	int value;
	int priority;
	Element *next; // in front
	Element *previous; // behind
};

struct Queue
{
	Element *head;
	Element *tail;
	int maxPriority;
};

Queue *createQueue()
{
	return new Queue{ nullptr };
}

bool isEmpty(Queue *queue)
{
	return queue->head == nullptr;
}

Queue *enqueue(Queue *queue, int value, int priority)
{
	if (priority < 0)
	{
		cout << "Error adding value with priority < 0" << endl;
		return queue;
	}

	Element *newElement = new Element{ value, priority, queue->tail, nullptr };
	if (isEmpty(queue))
	{
		queue->head = newElement;
		queue->tail = newElement;
		queue->maxPriority = priority;
	}
	else
	{
		Element *current = queue->head;
		while (current->previous != nullptr)
		{
			current = current->previous;
		}
		if (newElement->priority > queue->maxPriority)
		{
			queue->maxPriority = newElement->priority;
		}
		current->previous = newElement;
		newElement->next = current;
		queue->tail = newElement;
	}
	return queue;
}

int dequeue(Queue *queue)
{
	if (isEmpty(queue))
	{
		return -1;
	}
	
	Element *needed = queue->head;
	while (needed->priority != queue->maxPriority)
	{
		needed = needed->previous;
	}
	
	if ((needed == queue->tail) && (needed == queue->head))
	{
		int value = needed->value;
		delete needed;
		queue->head = nullptr;
		queue->tail = nullptr;
		queue->maxPriority = -1;
		return value;
	}
	else if (needed == queue->tail)
	{
		queue->tail = needed->next;
		needed->next->previous = nullptr;
	}
	else if (needed == queue->head)
	{
		queue->head = needed->previous;
		needed->previous->next = nullptr;
	}
	else
	{
		needed->next->previous = needed->previous;
		needed->previous->next = needed->next;
	}

	int maximumPriority = -1;
	Element *current = queue->head;
	while (current != nullptr)
	{
		if (maximumPriority < current->priority)
		{
			maximumPriority = current->priority;
		}
		current = current->previous;
	}
	queue->maxPriority = maximumPriority;

	int value = needed->value;
	delete needed;
	return value;
}

void deleteQueue(Queue *queue)
{
	if (!isEmpty(queue))
	{
		Element *current = queue->head;
		while (current != nullptr)
		{
			Element *temp = current;
			current = current->previous;
			delete temp;
		}
	}
	delete queue;
}
// Модуль должен иметь функцию enqueue(), принимающую на вход значение и численный приоритет, и функцию dequeue(), 
// возвращающую значение с наивысшим приоритетом и удаляющую его из очереди. Если очередь пуста, dequeue() должен возвращать -1.

File: Semester2/Test1/1.1/1.1/Queue.cs
using System;

namespace _1._1
{
    /// <summary>
    /// Priority queue's class
    /// </summary>
    public class Queue
    {
        private Node head = null;
        private Node tail = null;

        /// <summary>
        /// Node of the list
        /// </summary>
        private class Node
        {
            public string Value { get; }
            public int Priority { get; }
            public Node Next { get; set; }

            /// <summary>
            /// Node's constructor
            /// </summary>
            /// <param name="value">Node's value</param>
            public Node(string value, int priority)
            {
                this.Value = value;
                this.Priority = priority;
                Next = null;
            }
        }

        /// <summary>
        /// Checks if the queue is empty
        /// </summary>
        /// <returns>True if the queue is empty and false if it's not</returns>
        public bool IsEmpty() => head == null;

        /// <summary>
        /// Enqueues a value into the queue
        /// </summary>
        /// <param name="value">Value of given element</param>
        /// <param name="priority">Priority of given element</param>
        public void Enqueue(string value, int priority)
        {
            var newNode = new Node(value, priority);
            if (IsEmpty())
            {
                head = newNode;
                tail = newNode;
            }
            else
            {
                tail.Next = newNode;
                tail = newNode;
            }
        }

        /// <summary>
        /// Defines the value of highest priority in the queue
        /// </summary>
        /// <returns>Value of highest priority in the queue</returns>
        public int HighestPriority()
        {
            if (IsEmpty())
            {
                throw new EmptyQueueException("Queue is empty");
            }

            var current = head;
            var priority = head.Priority;
            while (current.Next != null)
            {
                if (priority < current.Priority)
                {
                    priority = current.Priority;
                }
                current = current.Next;
            }
            if (priority < current.Priority)
            {
                priority = current.Priority;
            }
            return priority;
        }

        /// <summary>
        /// Defines the position of element with highest priority in the queue
        /// </summary>
        /// <returns>Position of element with highest priority in the queue</returns>
        public int HighestPriorityPosition()
        {
            if (IsEmpty())
            {
                throw new EmptyQueueException("Queue is empty");
            }

            var current = head;
            var counter = 1;
            var maxPriority = HighestPriority();
            while (current.Priority != maxPriority)
            {
                current = current.Next;
                ++counter;
            }
            return counter;
        }

        /// <summary>
        /// Dequeues a value with highest priority
        /// </summary>
        /// <returns></returns>
        public string Dequeue()
        {
            if (IsEmpty())
            {
                throw new EmptyQueueException("Queue is empty");
            }

            var current = head;
            int position = HighestPriorityPosition();

            if (position == 1)
            {
                head = head.Next;
                return current.Value;
            }

            for (int i = 1; i < position - 1; ++i)
            {
                current = current.Next;
            }
            var value = current.Next.Value;
            current.Next = current.Next.Next;
            return value;
        }
    }
}

// Реализовать очередь с приоритетами. Очередь должна иметь метод Enqueue(), принимающий на вход значение и численный приоритет, 
// и метод Dequeue(), возвращающий значение с наивысшим приоритетом и удаляющий его из очереди. 
// Тип хранимых значений -- любой. Если очередь пуста, Dequeue() должен бросать исключение. Комментарии и юнит-тесты обязательны.